news/5126-samefile-regression

```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* <news item>
+
+### Bug fixes
+
+* Use `os.path.normpath` instead of `Path.samefile` to avoid `os.stat` calls. (#5126 via #5130)
```

Comment:
```suggestion
* Update `which_package` to use a cached mapping of files to packages (`O(1)`) instead of relying on `Path.samefile` comparisons (`O(n * m)`). (#5126 via #5130)
```